bftengine/src/bftengine/CryptoManager.cpp

```diff
@@ -0,0 +1,245 @@
+
+std::shared_ptr<CryptoManager> CryptoManager::init(std::unique_ptr<Cryptosystem>&& cryptoSys) {
+  ConcordAssert(cryptoSys != nullptr);
+  s_cm.reset(new CryptoManager(std::move(cryptoSys)));
+  LOG_INFO(s_cm->logger(), "CryptoManager Initialized");
+  return s_cm;
+}
+
+CryptoManager& CryptoManager::instance() {
+  ConcordAssertNE(s_cm.get(), nullptr);
+  return *s_cm;
+}
+std::shared_ptr<IThresholdSigner> CryptoManager::thresholdSignerForSlowPathCommit(const SeqNum sn) const {
+  return get(sn)->thresholdSigner_;
+}
+std::shared_ptr<IThresholdVerifier> CryptoManager::thresholdVerifierForSlowPathCommit(const SeqNum sn) const {
+  return get(sn)->thresholdVerifierForSlowPathCommit_;
+}
+std::shared_ptr<IThresholdSigner> CryptoManager::thresholdSignerForCommit(const SeqNum sn) const {
+  return get(sn)->thresholdSigner_;
+}
+std::shared_ptr<IThresholdVerifier> CryptoManager::thresholdVerifierForCommit(const SeqNum sn) const {
+  return get(sn)->thresholdVerifierForCommit_;
+}
+std::shared_ptr<IThresholdSigner> CryptoManager::thresholdSignerForOptimisticCommit(const SeqNum sn) const {
+  return get(sn)->thresholdSigner_;
+}
+std::shared_ptr<IThresholdVerifier> CryptoManager::thresholdVerifierForOptimisticCommit(const SeqNum sn) const {
+  return get(sn)->thresholdVerifierForOptimisticCommit_;
+}
+
+std::unique_ptr<Cryptosystem>& CryptoManager::getLatestCryptoSystem() const {
+  return checkpointToSystem().rbegin()->second->cryptosys_;
+}
+
+/**
+ * @return An algorithm identifier for the latest threshold signature scheme
+ */
+concord::crypto::SignatureAlgorithm CryptoManager::getLatestSignatureAlgorithm() const {
+  const std::unordered_map<std::string, concord::crypto::SignatureAlgorithm> typeToAlgorithm{
+      {MULTISIG_EDDSA_SCHEME, concord::crypto::SignatureAlgorithm::EdDSA},
+  };
+  auto currentType = getLatestCryptoSystem()->getType();
+  return typeToAlgorithm.at(currentType);
+}
+
+// IMultiSigKeyGenerator methods
+std::tuple<std::string, std::string, concord::crypto::SignatureAlgorithm> CryptoManager::generateMultisigKeyPair() {
+  LOG_INFO(logger(), "Generating new multisig key pair");
+  auto [priv, pub] = getLatestCryptoSystem()->generateNewKeyPair();
+  return {priv, pub, getLatestSignatureAlgorithm()};
+}
+
+void CryptoManager::syncPrivateKeyAfterST(const std::string& secretKey, const std::string& verificationKey) {
+  std::lock_guard<std::mutex> guard(mutex_);
+  for (auto& [checkpoint, cryptoSystem] : checkpointToSystem()) {
+    auto currentKey = cryptoSystem->cryptosys_->getMyVerificationKey();
+    LOG_INFO(logger(), "checking keys after ST" << KVLOG(checkpoint, currentKey, secretKey, verificationKey));
+    if (currentKey == verificationKey) {
+      cryptoSystem->cryptosys_->updateKeys(secretKey, verificationKey);
+      LOG_INFO(logger(), "Updated private key for cryptosystem with checkpoint:" << checkpoint);
+    }
+  }
+}
+
+void CryptoManager::onPrivateKeyExchange(const std::string& secretKey,
```

Comment:
onPrivateKeyRotation

Reply:
This is an implementation of the IKeyExchanger interface.
Since there are a lot of renames to do, I think it should be done in a separate PR.